src/FakeItEasy/Core/FakeManager.cs

```diff
@@ -219,5 +224,20 @@ private static void ApplyRule(CallRuleMetadata rule, IInterceptedFakeObjectCall
             rule.CalledNumberOfTimes++;
             rule.Rule.Apply(fakeObjectCall);
         }
+
+        private void UpdateLastSequenceNumber(int sequenceNumber)
+        {
+            //// Set the specified sequence number as the last sequence number if it's greater than the current last sequence number.
+            //// Because lastSequenceNumber might be changed by another thread after the comparison, we use CompareExchange to
+            //// only assign it if it has the same value as the one we compared with. If it's not the case, we retry.

```

Comment:
> `////`? I think we had this exchange once before, but why `////` instead of `//`? Is it because StyleCop complained about the blank line after the block with just `//`?

Yes

> Also, this comment does a good job of explaining what we're doing here, but I'd been hoping for a little more "'why".

Sure, I'll add more details

Reply:
Done

Reply:
>  how it ensured in-order insertion into the queue.

It doesn't. The calls could still be out of order, but it no longer affects whether the assertion succeeds or not. I'm not sure there's a way to ensure they're in order, because what *is* the right order? If `call1` is made on `thread1` before `call2` is made on `thread2`, `thread2` could still reach the point where a sequence number is assigned before `thread1`. The only way to prevent this would be to lock for the whole `Process` method. And even then, there would still be a race condition between when the calling code invokes the fake method and when the `Process` method is called.

Reply:
Sorry, yeah. I meant to say I wasted time trying to figure out why they were in order, when they weren't necessarily.

> what is the right order?

My take? The sequence number order. By definition. If calls are happening concurrently like that, then it's an arbitrary choice. I'm not proposing we change anything about how we handle things. Just conversation.

Reply:
> My take? The sequence number order. By definition.

I agree.

> I'm not proposing we change anything about how we handle things.

Maybe we should change one thing: order the calls by sequence number in the exception messages.

Reply:
> Maybe we should change one thing: order the calls by sequence number in the exception messages.

They already are for exceptions issued from failed ordered assertions, and maybe it makes less difference for unordered assertions, but we could do.